packages/console/src/foreign.ts

```diff
@@ -145,7 +146,8 @@ export class ForeignHandler implements IDisposable {
    * Create a new code cell for an input originated from a foreign session.
    */
   private _newCell(parentMsgId: string): CodeCell {
-    let cell = this._factory();
+    let cell = this.parent.createCodeCell();
+    cell.addClass(FOREIGN_CELL_CLASS);
     this._cells.set(parentMsgId, cell);
```

Comment:
This is the problem that I talked about. `_cells` is private to `ForeignHandler`, which is a map of `parentMsgId` and (reference to) `CodeCell` so the `parentMsgId` is only seen by the `ForeignHandler`, not by `console/widget` (whose `_cells` is a list of cells).

When both `ForeignHandler` and `TransientHandler` is enabled, the `ForeignHandler` will first get `execute_input` and create a cell, but `TransientHandler` does not know this cell has the same `parentMsgId` as the `transient_display_data` and will create a new cell.

So perhaps the mechanism of storing and querying `parentMsgId` need to be moved to console/widget.

Reply:
Another problem with keeping `_cells` in `ForeignHandler` is that if the core or another extension decides to remove some cells (e.g. to keep the use of memory in check if the console gets really long), the `ForeignHandler` will have invalid references in its `_cells`.

Reply:
@jasongrout Do you have an easy fix to the problem I mentioned or do you want me to work on a fix?

Reply:
Do you want to try out what I just pushed? Does that look better to you?

Reply:
Yes, it does look better. I will try it tonight by implementing my `TransientHandler` against it.

Reply:
I can confirm that the PR works perfectly for my purpose and can be merged as soon as the tests are fixed.

Reply:
Thanks again for patiently explaining the issues you saw. Very good points!